cardano-cli/test/cardano-cli-golden/Test/Golden/Byron/Tx.hs

```diff
@@ -53,16 +53,24 @@ hprop_byronTx = propertyOnce $ H.moduleWorkspace "tmp" $ \tempDir -> do
 
   compareByronTxs createdTx expectedTx
 
-getTxByteString :: FilePath -> H.PropertyT IO (ATxAux ByteString)
+getTxByteString :: ()
+  => Member (Embed IO) r
+  => Member (Error ByronTxError) r
+  => FilePath
+  -> Sem r (ATxAux ByteString)
 getTxByteString txFp = do
-  eATxAuxBS <- liftIO . runExceptT $ readByronTx $ File txFp
-  case eATxAuxBS of
-    Left err -> failWith Nothing . docToString $ renderByronTxError err
-    Right aTxAuxBS -> return aTxAuxBS
+  embed (runExceptT $ readByronTx $ File txFp)
+    & onLeftM throw
 
-compareByronTxs :: FilePath -> FilePath -> H.PropertyT IO ()
+compareByronTxs :: ()
+  => HasCallStack
+  => Member (Embed IO) r
+  => Member Hedgehog r
+  => FilePath
+  -> FilePath
+  -> Sem r ()
 compareByronTxs createdTx expectedTx = do
-  createdATxAuxBS <- getTxByteString createdTx
-  expectedATxAuxBS <- getTxByteString expectedTx
+  createdATxAuxBS <- getTxByteString createdTx & trapFail @ByronTxError
+  expectedATxAuxBS <- getTxByteString expectedTx & trapFail @ByronTxError
```

Comment:
What happens if we don't write those `trapFail` (where is this function defined btw? I could not find it)?

I'm worried we're not saving on boilerplate if for every IO call, we have to annotate it with `trapFail theExceptionThatItCanThrow`.

Reply:
What `trapFail` gives to is a quick way to catch the error and annotate it at the point of failure.

It is possible to `trapFail` once for the entire test, however what you get if you do that is the at the failure annotation will be at the `trapFail` at the catch-all level not at the operation that failed.

Reply:
The `trapFail` function is here: https://hackage.haskell.org/package/hw-polysemy

Reply:
> What happens if we don't write those trapFail?

If you don't call `trapFail` you get a compile error.  For example:

```
test/cardano-cli-golden/Test/Golden/Byron/Tx.hs:74:23: error: [GHC-39999]
    • Could not deduce ‘Member (Error ByronTxError) r’
        arising from a use of ‘getTxByteString’
      from the context: (HasCallStack, Member (Embed IO) r,
                         Member Hedgehog r)
        bound by the type signature for:
                   compareByronTxs :: forall (r :: Polysemy.Internal.Kind.EffectRow).
                                      (HasCallStack, Member (Embed IO) r, Member Hedgehog r) =>
                                      FilePath -> FilePath -> Sem r ()
        at test/cardano-cli-golden/Test/Golden/Byron/Tx.hs:(65,1)-(71,13)
    • In a stmt of a 'do' block:
        expectedATxAuxBS <- getTxByteString expectedTx
      In the expression:
        do createdATxAuxBS <- getTxByteString createdTx
                                & trapFail @ByronTxError
           expectedATxAuxBS <- getTxByteString expectedTx
           normalByronTxToGenTx expectedATxAuxBS
             === normalByronTxToGenTx createdATxAuxBS
      In an equation for ‘compareByronTxs’:
          compareByronTxs createdTx expectedTx
            = do createdATxAuxBS <- getTxByteString createdTx
                                      & trapFail @ByronTxError
                 expectedATxAuxBS <- getTxByteString expectedTx
                 normalByronTxToGenTx expectedATxAuxBS
                   === normalByronTxToGenTx createdATxAuxBS
   |
74 |   expectedATxAuxBS <- getTxByteString expectedTx
   |                       ^^^^^^^^^^^^^^^
```

Reply:
If you get an error that says `Could not deduce ‘Member (Error MyError) r’`, you have two choices.  You can handle it right there use `trap`, which is like `catch`, or `trapFail` which is like `trap` by fails immediately with a hedgehog annotation or you can propagate the error by adding `Member (Error MyError) r` to the function signature in which case you punt the responsibility of handling the error to the caller.

Note that a function can declare that it throws multiple errors so you don't have to create a new sum type with many constructors to propagate multiple errors.